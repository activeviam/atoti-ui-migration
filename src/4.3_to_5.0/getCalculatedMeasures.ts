import { ContentRecord } from "@activeviam/activeui-sdk-5.0";
import xml2js from "xml2js";
import _flatMap from "lodash/flatMap";
import { LegacyCalculatedMeasure } from "./migrateCalculatedMeasures";

/**
 * Extracts and parses the XML calculated measure objects from the /pivot/entitlements/cm folder.
 */
export const getCalculatedMeasures = async (
  calculatedMeasuresFolder: ContentRecord,
): Promise<LegacyCalculatedMeasure[]> => {
  const parser = new xml2js.Parser();

  const calculatedMeasures: LegacyCalculatedMeasure[] = [];

  if (calculatedMeasuresFolder?.children) {
    await Promise.all(
      // Legacy calculated measures are grouped by cube in `/pivot/entitlements/cm`. But they are flattened in Atoti UI 5.
      _flatMap(calculatedMeasuresFolder.children, async ({ children }) => {
        if (children) {
          const measures = await Promise.all(
            Object.values(children).map(async (cmEntry) => {
              const result = await parser.parseStringPromise(
                cmEntry.entry.content,
              );

              const { owners, readers } = cmEntry.entry;
              /*
                  xml2js parses the given XML and returns an object of the shape:

                  {
                    calculatedMember: {$: {...}, additionalProperties: {...}}
                  }

                  In this case, the data that is required (`expression`, `formatStringExpression`, and `uniqueName`) lives within the `$` property.
                */
              const { expression, formatStringExpression, uniqueName } =
                result.calculatedMember["$"];

              const calculatedMeasure = {
                owners,
                readers,
                expression,
                formatStringExpression,
                uniqueName,
              };

              return calculatedMeasure;
            }),
          );
          calculatedMeasures.push(...measures);
        }
      }),
    );
  }

  return calculatedMeasures;
};
