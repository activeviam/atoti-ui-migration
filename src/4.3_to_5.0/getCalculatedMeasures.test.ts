import { getCalculatedMeasures } from "./getCalculatedMeasures";

describe("getCalculatedMeasures", () => {
  it("returns an empty array when the calculated measures folder contains no children", async () => {
    const result = await getCalculatedMeasures({
      entry: {
        owners: ["ROLE_CS_ROOT"],
        readers: ["ROLE_CS_ROOT"],
        isDirectory: true,
      },
    });
    expect(result).toEqual([]);
  });

  it("returns the parsed calculated measure objects", async () => {
    const legacyCalculatedMeasuresFolder = {
      entry: {
        isDirectory: true,
        owners: ["ROLE_CS_ROOT"],
        readers: ["ROLE_CS_ROOT"],
      },
      children: {
        EquityDerivativesCube: {
          entry: {
            isDirectory: true,
            owners: ["ROLE_CS_ROOT"],
            readers: ["ROLE_CS_ROOT"],
          },
          children: {
            "[Measures].[second calculated measure]": {
              entry: {
                content:
                  '<?xml version="1.0" encoding="UTF-8" standalone="yes"?>\n<calculatedMember expression="IIf(IsEmpty(([Currency].[Currency].[ALL].[AllMember], [Measures].[pnl.FOREX])), null, [Measures].[pnl.FOREX] / ([Currency].[Currency].[ALL].[AllMember], [Measures].[pnl.FOREX]))" formatStringExpression="&quot;#,###.##%&quot;" uniqueName="[Measures].[second calculated measure]" xmlns="http://www.quartetfs.com">\n    <additionalProperties/>\n</calculatedMember>\n',
                isDirectory: false,
                owners: ["admin"],
                readers: ["admin"],
              },
            },
            "[Measures].[first calculated measure]": {
              entry: {
                content:
                  '<?xml version="1.0" encoding="UTF-8" standalone="yes"?>\n<calculatedMember expression="Count(Descendants([CounterParty].[CounterParty].CurrentMember, [CounterParty].[CounterParty].[CounterPartyGroup]), EXCLUDEEMPTY)" formatStringExpression="&quot;#,###.##&quot;" uniqueName="[Measures].[first calculated measure]" xmlns="http://www.quartetfs.com">\n    <additionalProperties/>\n</calculatedMember>\n',
                isDirectory: false,
                owners: ["admin"],
                readers: ["admin"],
              },
            },
          },
        },
      },
    };
    const result = await getCalculatedMeasures(legacyCalculatedMeasuresFolder);

    expect(result).toMatchSnapshot();
  });

  // https://activeviam.atlassian.net/browse/UI-8803
  it("parses an XML calculated measure with a namespace", async () => {
    // The "calculateMember" tag has a namespace: "ns2".
    const legacyCalculatedMeasuresFolder = {
      entry: {
        isDirectory: true,
        owners: ["ROLE_CS_ROOT"],
        readers: ["ROLE_CS_ROOT"],
      },
      children: {
        EquityDerivativesCube: {
          entry: {
            isDirectory: true,
            owners: ["ROLE_CS_ROOT"],
            readers: ["ROLE_CS_ROOT"],
          },

          children: {
            "[Measures].[CC Impact MtM]": {
              entry: {
                content:
                  '<?xml version="1.0" encoding="UTF-8" standalone="yes"?>\n<ns2:calculatedMember expression="[Measures].[deltaMTM]" formatStringExpression="&quot;#,###,,&quot;" uniqueName="[Measures].[CC Impact MtM]" xmlns:ns2="http://www.quartetfs.com">\n    <ns2:additionalProperties/>\n</ns2:calculatedMember>\n',
                isDirectory: false,
                owners: ["43890025"],
                readers: ["43890025"],
                timestamp: 1569424639841,
                lastEditor: "43890025",
                canRead: true,
                canWrite: true,
              },
            },
          },
        },
      },
    };

    const result = await getCalculatedMeasures(legacyCalculatedMeasuresFolder);

    expect(result).toMatchSnapshot();
  });
});
