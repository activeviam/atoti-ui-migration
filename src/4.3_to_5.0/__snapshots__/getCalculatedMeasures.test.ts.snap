// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getCalculatedMeasures returns the parsed calculated measure objects 1`] = `
[
  {
    "expression": "IIf(IsEmpty(([Currency].[Currency].[ALL].[AllMember], [Measures].[pnl.FOREX])), null, [Measures].[pnl.FOREX] / ([Currency].[Currency].[ALL].[AllMember], [Measures].[pnl.FOREX]))",
    "formatStringExpression": ""#,###.##%"",
    "owners": [
      "admin",
    ],
    "readers": [
      "admin",
    ],
    "uniqueName": "[Measures].[second calculated measure]",
  },
  {
    "expression": "Count(Descendants([CounterParty].[CounterParty].CurrentMember, [CounterParty].[CounterParty].[CounterPartyGroup]), EXCLUDEEMPTY)",
    "formatStringExpression": ""#,###.##"",
    "owners": [
      "admin",
    ],
    "readers": [
      "admin",
    ],
    "uniqueName": "[Measures].[first calculated measure]",
  },
]
`;
